### PR TITLE
Add asmjs/wasm32 to CI

### PR DESCRIPTION
**Pull request by alexcrichton**
Rebase of #610 and also move emscripten up much higher in the hierarchy to
ensure that it doesn't have too much of a ripple effect on other platforms.

This involved moving down a good number of definitions, but hopefully was done
with care to not break anything!

### CODE REVIEW

File: ci/emscripten.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+cd /
+curl -L https://s3.amazonaws.com/mozilla-games/emscripten/releases/emsdk-portable.tar.gz | \
+    tar -xz
+
+cd /emsdk-portable
+./emsdk update
+hide_output ./emsdk install sdk-1.37.14-64bit
+./emsdk activate sdk-1.37.14-64bit
```
Comment (lilianmoraru) [inline ci/emscripten.sh]: Could you change these please to use the latest current release[ `1.37.20` ]?: https://github.com/kripken/emscripten/releases

**Comment by lilianmoraru** [comment]
I searched for `target_env = "musl"` and I am curious if these don't need also `target_os = "emscripten"`(while the merge request was open, other structures were added): [1](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/notbsd/mod.rs#L123), [2](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/mod.rs#L675), [3](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/mod.rs#L709), [4](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/notbsd/linux/mod.rs#L109), [5](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/notbsd/linux/mod.rs#L117), [6](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/notbsd/linux/mod.rs#L1375)

And [no doc links](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/lib.rs#L36) for `wasm32` and `emscripten`

**Comment by lilianmoraru** [comment]
- 1 seems [ok](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/include/libc/sys/epoll.h#L46)?
- 2 also [ok](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/lib/libc/musl/src/string/strerror_r.c#L5)?(seems weird since it should be basically `musl`, as the `emscripten` commits also suggest)
- 3 [not ok](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/include/libc/resolv.h#L128)?(not in the `cfg` in general)
- 4 and 5 [not ok](https://github.com/kripken/emscripten/blob/6dc4ac5f9e4d8484e273e4dcc554f809738cedd6/system/lib/libc/musl/arch/emscripten/bits/alltypes.h#L367)?([the other](https://github.com/alexcrichton/libc/blob/22b98dedfee82bd8e08d979093eaca355ab85e9c/src/unix/notbsd/linux/mod.rs#L96) `cfg`s seem to ignore `musl` also)
- 6 probably not ok because `emscripten` is basically `musl`.

**Comment by alexcrichton** [comment]
@lilianmoraru 

> I searched for target_env = "musl" and I am curious if these don't need also target_os = "emscripten"(while the merge request was open, other structures were added): 

Due to the way CI works we shouldn't need to worry about any of these, as CI would fail if any definition were inocorrect.

**Comment by lilianmoraru** [comment]
@alexcrichton You repeated the merge/rebase? That must of been painful :)
Thanks!

**Comment by lilianmoraru** [comment]
@alexcrichton Big thanks for finishing this up!

**Pull request merged**